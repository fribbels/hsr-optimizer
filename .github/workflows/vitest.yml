name: Vitest Tests
on:
  push:
    branches: [ main, beta ]
  pull_request:
    branches: [ main, beta ]
    types: [ opened, reopened, synchronize, review_requested, ready_for_review ]
jobs:
  test:
    if: github.event.pull_request.draft == false
    timeout-minutes: 10
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-node@v4
        with:
          node-version: 22
      - name: Install dependencies
        run: npm ci
      - name: Run Vitest tests
        run: npm run vitest