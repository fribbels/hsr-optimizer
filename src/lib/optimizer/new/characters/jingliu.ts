import { EarlyConditional } from '../stats/conditional'
import { HsrElement, Trait } from '../stats/context'
import { matchByTraits } from '../stats/matcher'
import { Modifiers } from '../stats/modifier'
import { StepBuilder } from '../step/builder'
import { Character, CharacterPreset, CharacterSteps, Eidolon } from './character'

type JingliuMetadata = {
  talentEnhancedState: boolean
  talentHpDrainAtkBuff: number
  e1CdBuff: boolean
  e2SkillDmgBuff: boolean
}

const DEFAULT_METADATA = {
  talentEnhancedState: true,
  talentHpDrainAtkBuff: 1.8,
  e1CdBuff: true,
  e2SkillDmgBuff: true,
}
export class Jingliu extends Character<JingliuMetadata> {
  private modifiers: Modifiers = {
    early: [],
    late: [],
    unconditional: [],
  }

  constructor(eid: Eidolon, metadata: JingliuMetadata = DEFAULT_METADATA) {
    super(eid, metadata)
    if (eid > 0 && metadata.e1CdBuff) {
      this.modifiers.unconditional.push({ crit: { critDmg: 0.25 } })
    }
    if (eid > 1 && metadata.e2SkillDmgBuff) {
      this.modifiers.early.push(new EarlyConditional(matchByTraits(Trait.SKILL), { dmgBoost: 0.8 }))
    }
    if (metadata.talentEnhancedState) {
      this.modifiers.unconditional.push({ basic: { percent: { atk: metadata.talentHpDrainAtkBuff } } })
    }
  }

  asTeammate: Modifiers = this.modifiers

  asOptiTarget: CharacterSteps[] = [
    {
      type: 'skill',
      modifiers: this.modifiers,
      // It is possible to handle hitsplit here, should we?
      steps: [StepBuilder.damage(HsrElement.ICE, [Trait.SKILL], 2.5, 'atk')],
    },
  ]

  presets: CharacterPreset[] = []
}
