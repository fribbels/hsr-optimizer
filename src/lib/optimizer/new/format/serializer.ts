import { Serializable } from './serializable'

/* eslint-disable @typescript-eslint/ban-types */
type Constructor = Function

type ClassInstance = {
  constructor: Constructor
}

type SerializationOptions = {
  /**
   * Making sure that this {@link ObjectMapper} can deserialize the result of
   * this {@link ObjectMapper.serialize serialization}
   */
  collectMetadata: boolean
}

type DeserializationOptions = {
  /**
   * Use ```eval(className)``` as a last resort when deserializing unknown class
   * name.
   *
   * IMPORTANT: If you need to use this, something is very wrong.
   */
  useEvalAsFallback: boolean
}

type Deserializable = {
  /**
   * An unique name that identify a Serializable. This value is used to map the
   * serialized representation to the corresponding class form. This should be
   * ```class.constructor.name```
   */
  __serializable_name: SerializableName
}

// A type hack so that a string is not mistakenly used in the wrong context
type SerializableName = string & { __ignored: undefined }

/**
 * The primary API for serialization and deserialization.
 */
export class ObjectMapper {
  private map: Map<SerializableName, Function> = new Map()
  /**
   * @param classes classes that are supported for deserialization
   */
  constructor(classes: Constructor[] = []) {
    classes.forEach((clazz) => populateMap(clazz, this.map))
  }

  /**
   * Serialize an instance of any type to a string that can be
   * ```JSON.parse()``` into a {@link Deserializable}.
   */
  serialize(instance: unknown, options?: Partial<SerializationOptions>): string {
    return JSON.stringify(instance, (_key, value) => {
      if (!isClassInstance(value)) {
        return value as unknown
      }
      if (options?.collectMetadata && !this.map.has(value.constructor.name as SerializableName)) {
        populateMap(value.constructor, this.map)
      }
      return Object.assign(isSerializable(value) ? value.serialize() : value, {
        __serializable_name: value.constructor.name,
      })
    })
  }

  deserialize<T>(json: string, options?: Partial<DeserializationOptions>): T {
    if (options?.useEvalAsFallback) {
      console.log(`
You are using eval() as a fallback option to find the deserializers for unknown classes.
If you don't understand the above warning (an end user), you are likely a victim of an attack.
I don't know how you ended up here as this is an internal library, but don't listen to whatever they told you. 

Even if it is intentional, this is a huge security risk, and probably dodgy programming. 
Are you sure this is neccessary? I bet it's not.
This feature is only useful for development.
Try to provide the full range of deserializable classes before deserialization.

I'll be honest with you, I did not even tested this feature. 
Whether this works or not is up to luck.`)
    }
    return JSON.parse(json, (_key, value) => {
      if (isDeserializable(value)) {
        return this.__deserialize(value, options?.useEvalAsFallback ?? false)
      }
      return value as unknown
    }) as T
  }

  private __deserialize(json: Deserializable, useEval: boolean): unknown {
    const deserializer = this.map.get(json.__serializable_name)
    if (!deserializer) {
      const errString = `
Can't locate deserializer in the map. Context:
__serializable_name (class name): ${json.__serializable_name}
Map: ${JSON.stringify(Array.from(this.map.entries()), null, 4)}
Can this ObjectMapper support this class?`
      if (!useEval) {
        throw new Error(errString)
      }

      return evalUnknownClass(json, errString)(json)
    }
    return deserializer(json)
  }

  /**
   * Making sure that this {@link ObjectMapper} can deserialize the value of the
   * given class.
   *
   * Note: it isn't aware of the nested properties' classes.
   * Javascript is a very dynamic language, thus at runtime there is no type
   * information for a specific property of a class. Can't be helped. You can
   * query that a class has a property with name ```fieldxxx```, but that field
   * can hold anything from a number, a string to a class instance.
   */
  supportClasses(...clazzes: Constructor[]) {
    // One such way to provide the type information at runtime is through the
    // stage 3 decorator proposal, which is what TypedJSON is doing.
    clazzes.forEach((clazz) => populateMap(clazz, this.map))
  }

  /**
   * Making sure that this {@link ObjectMapper} is aware of the given object
   * structure (for deserialization). There is very little reason to use this
   * method.
   */
  support(value: ClassInstance | object) {
    if (typeof value !== 'object') {
      console.log(`Can't support function: ${String(value)}`)
    }
    if (isClassInstance(value)) {
      if (this.map.has(value.constructor.name as SerializableName)) {
        return
      }
      console.log(`Supporting class ${value.constructor.name}`)

      populateMap(value.constructor, this.map)
    }
    // support all nested properties
    Object.values(value).filter((val): val is object => typeof val === 'object').forEach((val) => this.support(val))
  }

  isSupported(name: string) {
    return this.map.has(name as SerializableName)
  }
}

// ----------------
// HELPER FUNCTIONS
// ----------------
// Note: clazz is in fact the constructor function
export function populateMap(clazz: Constructor, map: Map<SerializableName, Function>) {
  let deserializer: Function
  if (isSerializable(clazz.prototype)) {
    deserializer = clazz.prototype.__deserialize.bind(undefined)
  } else {
    deserializer = function(json: unknown) {
      return Object.assign(Object.create(clazz.prototype as object), json) as unknown
    }
  }
  return map.set(
    clazz.name as SerializableName,
    deserializer,
  )
}

export function isClassInstance(value: unknown): value is ClassInstance {
  return value !== undefined
    && typeof value !== 'function'
    && value?.constructor?.name !== 'Object'
    && value?.constructor?.name !== 'Array'
}

export function isDeserializable(value: unknown): value is Deserializable {
  return (value as Deserializable)?.__serializable_name !== undefined
}

export function isSerializable(value: unknown): value is Serializable<object, object> {
  return (value as Serializable<object, object>)?.__deserialize !== undefined
}

// Least dodgy programming practice
export function evalUnknownClass(json: Deserializable, errString: string) {
  const clazz = eval(json.__serializable_name) as unknown
  if (typeof clazz !== 'function') {
    throw new Error(
      errString + `
Basically eval failed lmao. 
The environment (closures) doesn't have any class with the given name.
Web worker specific: Is the class referenced in your web worker? 
Vite would be smart enough to bundle that class into the web worker bundle if you do so.
If you indeed did so, it is likely that the class is shadowed in a closure.`,
    )
  }
  if (!clazz.toString().startsWith('class')) {
    throw new Error(
      errString + `
Eval results in a function, but it is not a class declaration.
Is the class declaration shadowed in a closure?`,
    )
  }
  // well, for better or worse, we did got a class constructor
  return function(json: unknown) {
    return Object.assign(Object.create(clazz.prototype as object), json) as unknown
  }
}
