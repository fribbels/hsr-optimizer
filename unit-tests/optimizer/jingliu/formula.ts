import { HsrElement, Trait } from 'lib/optimizer/new/stats/context'
import { StepBuilder } from 'lib/optimizer/new/step/builder'
import { Formula } from 'lib/optimizer/new/step/formula'

const unconditional = [
  {
    // 3 stack LC
    dmgBoost: 0.42,
    // LC passive
    crit: { critDmg: 0.2 },
    targetDef: { percent: -0.12 },
  },
  {
    // Talent - 180% ATK and 50% CR
    crit: { critRate: 0.5 },
    basic: { percent: { atk: 1.8 } },
  },
  {
    // Relevant traces - 37.3% Crit DMG
    crit: { critDmg: 0.373 },
    basic: {
      flat: {
        speed: 9,
      },
    },
  },
  {
    // Character base Crit
    crit: {
      critRate: 0.05,
      critDmg: 0.5,
    },
  },
]

const context = {
  basic: {
    lv: 80,
    base: {
      hp: 9999,
      atk: 1261.3,
      def: 9999,
      speed: 96,
    },
  },
  maxEnergy: 9999,
  targetBaseDef: 200 + 10 * 95,
  baseMods: {
    unconditional: unconditional,
    early: [],
    late: [],
  },
}

export const formula: Formula = Formula.create(
  [StepBuilder.damage(HsrElement.ICE, [Trait.SKILL], 2.5, 'atk').build()],
  context,
)

export const formulaWithLimit: Formula = Formula.create(
  [
    StepBuilder
      .damage(HsrElement.ICE, [Trait.SKILL], 2.5, 'atk')
      .limit({ basic: { speed: { min: 133 } } })
      .build(),
  ],
  context,
)
