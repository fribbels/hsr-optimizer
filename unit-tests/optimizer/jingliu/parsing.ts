import { HsrElement } from 'lib/optimizer/new/stats/context'
import { GeneralSubStats, SpherePiece } from 'lib/optimizer/new/stats/relic'

/**
 * DON'T COPY THE RELIC PARSING OF THIS FILE, IT'S NOT THE BEST WAY
 */

export type Relic = {
  part: string
  set: string
  main: {
    stat: string
    value: number
  }
  substats: {
    stat: string
    value: number
  }[]
}

export function parseRelic<T>(relic: Relic): T {
  return parseSubStats({ set: relic.set }, relic.substats.concat(relic.main)) as T
}

export function parseSubStats<T extends GeneralSubStats>(
  obj: T,
  subs: { stat: string; value: number }[],
) {
  subs.forEach((sub) => parseSub(obj, sub))
  return obj
}

export function parseSub<T extends GeneralSubStats>(
  obj: T,
  sub: { stat: string; value: number },
) {
  if (sub.stat === 'CRIT Rate') {
    safeAssign(obj, sub.value / 100, 'crit', 'critRate')
  } else if (sub.stat === 'CRIT DMG') {
    safeAssign(obj, sub.value / 100, 'crit', 'critDmg')
  } else if (sub.stat === 'Effect RES') {
    safeAssign(obj, sub.value, 'effectRes')
  } else if (sub.stat === 'Effect Hit Rate') {
    safeAssign(obj, sub.value, 'effectHitRate')
  } else if (sub.stat === 'HP%') {
    safeAssign(obj, sub.value / 100, 'basic', 'percent', 'hp')
  } else if (sub.stat === 'ATK%') {
    safeAssign(obj, sub.value / 100, 'basic', 'percent', 'atk')
  } else if (sub.stat === 'DEF%') {
    safeAssign(obj, sub.value / 100, 'basic', 'percent', 'def')
  } else if (sub.stat === 'SPD') {
    safeAssign(obj, sub.value, 'basic', 'flat', 'speed')
  } else if (sub.stat === 'HP') {
    safeAssign(obj, sub.value, 'basic', 'flat', 'hp')
  } else if (sub.stat === 'ATK') {
    safeAssign(obj, sub.value, 'basic', 'flat', 'atk')
  } else if (sub.stat === 'DEF') {
    safeAssign(obj, sub.value, 'basic', 'flat', 'def')
  } else if (sub.stat === 'Energy Regeneration Rate') {
    obj['energyRegenerationRate'] = sub.value / 100
  } else if (sub.stat === 'Break Effect') {
    safeAssign(obj, sub.value / 100, 'breakEffect')
  } else if (sub.stat === 'Outgoing Healing Boost') {
    obj['outgoingHealing'] = sub.value / 100
  } else if (sub.stat.includes('DMG Boost')) {
    ;(obj as SpherePiece).__dmgBoost = {
      ele: HsrElement[sub.stat.split(' ')[0].toUpperCase()],
      value: sub.value / 100,
    }
  }
  return obj
}

function safeAssign(obj: object, val: number, ...keys: string[]) {
  let o = obj
  for (let i = 0; i < keys.length - 1; i++) {
    const key = keys[i]
    if (!o[key]) {
      o[key] = {}
    }
    o = o[key]
  }
  const last = keys[keys.length - 1]
  if (o[last]) {
    o[last] += val
  } else {
    o[last] = val
  }
}
