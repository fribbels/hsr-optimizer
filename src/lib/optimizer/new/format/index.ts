import { BasicPercentageStats } from '../stats/basicStat'
import { EarlyConditional, FixedStat, LateConditional, TransformingStat } from '../stats/conditional'
import {
  AllMatcher,
  AlwaysMatcher,
  AnyMatcher,
  EarlyAllMatcher,
  EarlyAnyMatcher,
  ElementMatcher,
  TraitMatcher,
  UnassumingStatMatcher,
} from '../stats/matcher'
import { StatCollector } from '../stats/stat'
import { StatBuilder } from '../stats/statBuilder'
import { Formula } from '../step/formula'
import { NormalDamageStep } from '../step/step'
import { ObjectMapper } from './serializer'
/**
 * So what is this class?
 *
 * Short answer: A serialization/deserialization mini library.
 *
 * Long answer: TODO let me collect my thought before I create a medium account lmao
 */
const DEFAULT_CLASSES = [
  BasicPercentageStats,
  StatCollector,
  StatBuilder,
  Formula,
  NormalDamageStep,
  // --------
  // Matchers
  // --------
  ElementMatcher,
  TraitMatcher,
  UnassumingStatMatcher,
  AllMatcher,
  EarlyAllMatcher,
  AnyMatcher,
  EarlyAnyMatcher,
  AlwaysMatcher,
  // ------------
  // Conditionals
  // ------------
  EarlyConditional,
  LateConditional,
  FixedStat,
  TransformingStat,
]

const DEFAULT_OBJECT_MAPPER = new ObjectMapper(DEFAULT_CLASSES)

export function serialize(instance: unknown): string {
  return DEFAULT_OBJECT_MAPPER.serialize(instance)
}

export function deserialize<T>(json: string): T {
  return DEFAULT_OBJECT_MAPPER.deserialize(json)
}
