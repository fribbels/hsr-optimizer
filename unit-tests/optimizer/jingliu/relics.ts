import sample from 'data/sample-save.json'
import { EarlyConditional, FixedStat, LateConditional } from 'lib/optimizer/new/stats/conditional'
import { HsrElement, SupportedContextStat, Trait } from 'lib/optimizer/new/stats/context'
import {
  alwaysMatch,
  matchAll,
  matchByElement,
  matchByStat,
  matchByTraits,
  SupportedMatchType,
} from 'lib/optimizer/new/stats/matcher'
import {
  BodyPiece,
  FeetPiece,
  HandPiece,
  HeadPiece,
  RelicSetEffect,
  RopePiece,
  SpherePiece,
} from 'lib/optimizer/new/stats/relic'
import { parseRelic, Relic } from './parsing'

const unparsedRelics: Relic[] = sample.relics

const HUNTER_OF_GLACIAL_FOREST = 'Hunter of Glacial Forest'
const hunterOfGlacialForest: RelicSetEffect = {
  set2: {
    // 10% Ice DMG Bonus
    early: [new EarlyConditional(matchByElement(HsrElement.ICE), { dmgBoost: 0.1 })],
    late: [],
  },
  set4: {
    // 25% Crit DMG
    early: [new EarlyConditional(alwaysMatch(), { crit: { critDmg: 0.25 } })],
    late: [],
  },
}

const RUTILANT_ARENA = 'Rutilant Arena'
const rutilantArena = {
  set2: {
    early: [new EarlyConditional(alwaysMatch(), { crit: { critRate: 0.08 } })],
    late: [
      new LateConditional(
        // Normal ATK and Skill DMG Boost if Crit Rate >= 0.7
        matchAll(
          matchByTraits(Trait.SKILL, Trait.NORMAL),
          matchByStat(SupportedContextStat.CRIT_RATE, SupportedMatchType.GREATER_THAN_OR_EQUAL, 0.7),
        ),
        new FixedStat({ dmgBoost: 0.2 }),
      ),
    ],
  },
}

const GENIUS_OF_BRILLIANT_STARS = 'Genius of Brilliant Stars'
const geniusOfBrilliantStars = {
  set2: {
    // 10% Quantum DMG Bonus
    early: [new EarlyConditional(matchByElement(HsrElement.QUANTUM), { dmgBoost: 0.1 })],
    late: [],
  },
  set4: {
    // 20% DEF PEN
    early: [new EarlyConditional(alwaysMatch(), { targetDef: { percent: -0.2 } })],
    late: [],
  },
}

const ALL_CONSIDERED_SETS = [HUNTER_OF_GLACIAL_FOREST, RUTILANT_ARENA, GENIUS_OF_BRILLIANT_STARS]
const LIMITED_SETS = [HUNTER_OF_GLACIAL_FOREST, RUTILANT_ARENA]

type RelicMap = {
  head: HeadPiece[]
  hand: HandPiece[]
  body: BodyPiece[]
  feet: FeetPiece[]
  sphere: SpherePiece[]
  rope: RopePiece[]
}

function reducerFor(sets: string[]) {
  return (map: RelicMap, relic: Relic) => {
    if (!sets.includes(relic.set)) {
      return map
    }
    switch (relic.part) {
      case 'Head':
        map.head.push(parseRelic(relic))
        break
      case 'Hands':
        map.hand.push(parseRelic(relic))
        break
      case 'Body':
        map.body.push(parseRelic(relic))
        break
      case 'Feet':
        map.feet.push(parseRelic(relic))
        break
      case 'PlanarSphere':
        map.sphere.push(parseRelic(relic))
        break
      case 'LinkRope':
        map.rope.push(parseRelic(relic))
        break
      default:
        throw new Error(`Unknown relic part ${relic.part}`)
    }
    return map
  }
}

const map = unparsedRelics.reduce<RelicMap>(reducerFor(ALL_CONSIDERED_SETS), {
  head: [],
  hand: [],
  body: [],
  feet: [],
  sphere: [],
  rope: [],
})

export const extendedRelics: RelicMap = map

export const allSetEffects: { [K: string]: RelicSetEffect } = {
  [HUNTER_OF_GLACIAL_FOREST]: hunterOfGlacialForest,
  [RUTILANT_ARENA]: rutilantArena,
  [GENIUS_OF_BRILLIANT_STARS]: geniusOfBrilliantStars,
}

export const limitedRelics = unparsedRelics.reduce(reducerFor(LIMITED_SETS), {
  head: [],
  hand: [],
  body: [],
  feet: [],
  sphere: [],
  rope: [],
})

export const limitedSetEffects = {
  [HUNTER_OF_GLACIAL_FOREST]: hunterOfGlacialForest,
  [RUTILANT_ARENA]: rutilantArena,
}
