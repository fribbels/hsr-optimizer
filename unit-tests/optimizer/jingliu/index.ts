import { OptimizationRequest } from 'lib/optimizer/new/request'
import { formula, formulaWithLimit } from './formula'
import { allSetEffects, extendedRelics, limitedRelics, limitedSetEffects } from './relics'

const limited: OptimizationRequest = {
  relics: {
    pieces: limitedRelics,
    sets: limitedSetEffects,
  },
  formula: formula,
}

const extended: OptimizationRequest = {
  relics: {
    pieces: extendedRelics,
    sets: allSetEffects,
  },
  formula: formula,
}

const extendedWithLimit: OptimizationRequest = {
  relics: {
    pieces: extendedRelics,
    sets: allSetEffects,
  },
  formula: formulaWithLimit,
}
export { extended, extendedWithLimit, limited }
