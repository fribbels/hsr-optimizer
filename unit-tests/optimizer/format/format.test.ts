import { deserialize, serialize } from 'lib/optimizer/new/format'
import { ObjectMapper } from 'lib/optimizer/new/format/serializer'
import { OptimizationRequest } from 'lib/optimizer/new/request'
import { BasicPercentageStats } from 'lib/optimizer/new/stats/basicStat'
import { Trait } from 'lib/optimizer/new/stats/context'
import { StatCollector } from 'lib/optimizer/new/stats/stat'
import { describe, expect, test } from 'vitest'
import { limited } from '../jingliu'

describe('serialization', () => {

  test('basicPercentageClass', () => {
    const zero = new BasicPercentageStats(90, {
      atk: 900,
      hp: 800,
      def: 700,
      speed: 600,
    })

    expect(JSON.stringify(JSON.parse(serialize(zero)), null, 4), 'serialization should be stable').toMatchSnapshot()

    expect(zero.atk).toEqual(deserialize<BasicPercentageStats>(serialize(zero)).atk)
  })

  test('statCollector', () => {
    const zero = StatCollector.zero({
      basic: {
        lv: 90,
        base: {
          atk: 900,
          hp: 800,
          def: 700,
          speed: 600,
        },
      },
      traits: [Trait.DOT_BLEED, Trait.FOLLOW_UP, Trait.NORMAL],
      maxEnergy: 120,
      targetBaseDef: 200 + 10 * 95,
    })
    zero.add({
      basic: {
        percent: {
          atk: 0.37,
        },
      },
    })

    expect(serialize(zero), 'serialized form should contain class name metadata').toContain('__serializable_name')

    expect(zero, 'serialization should be invariant')
      .toMatchObject<StatCollector>(deserialize(serialize(zero)))
  })

  test('Optimization Request', () => {
    const optimizationRequest = limited
    const newReq = deserialize<OptimizationRequest>(structuredClone(serialize(optimizationRequest)))
    expect(
      newReq.formula.calculate([], [], []),
      `optimization request should be safe passing through structuredClone`,
    ).toMatchObject(optimizationRequest.formula.calculate([], [], []))
  })

  test('ObjectMapper Class', () => {
    const mapper = new ObjectMapper()
    mapper.support(limited)
    const serialized = mapper.serialize(limited)

    expect(serialized, 'ObjectMapper support should works with deserialization').toMatchObject(serialize(limited))
  })
})
