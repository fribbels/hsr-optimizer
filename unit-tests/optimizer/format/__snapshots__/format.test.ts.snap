// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`serialization > basicPercentageClass 1`] = `
"{
    "lv": 90,
    "base": {
        "atk": 900,
        "hp": 800,
        "def": 700,
        "speed": 600
    },
    "flat": {
        "atk": 0,
        "def": 0,
        "hp": 0,
        "speed": 0
    },
    "percent": {
        "atk": 0,
        "def": 0,
        "hp": 0,
        "speed": 0
    },
    "__serializable_name": "BasicPercentageStats"
}"
`;
